### PR TITLE
Restore date aggregation performance in UTC case

### LINKED ISSUE #37826: Drop in indexing performance and increase in agg latency after mappings/aggs switch to java time in #36363
#36363 (merged to master in https://github.com/elastic/elasticsearch/commit/daa2ec8a605d385a65b9ab3e89d016b3fd0dffe2) has caused a performance drop manifested in both a drop in indexing throughput (see: https://elasticsearch-benchmarks.elastic.co/#tracks/nyc-taxis/nightly/30d for `http_logs` / `nyc_taxis` / `nested` / `noaa`) as well as latency increase in aggs (see: `nyc_taxis`, `nested`, `http_logs`) and `gc` (`nested`) on 2018-01-23.

- Before: `nyc_taxis` median indexing throughput: 81421 docs/s 
- After: `nyc_taxis` median indexing throughput: 65981 docs/s

### Reproduction steps:

Java time commit `daa2ec8`:

```
esrally --pipeline="from-sources-complete" --track="nyc_taxis" --car="4gheap" --runtime-jdk=8 --challenge="append-no-conflicts" --revision="daa2ec8" --include-tasks="delete-index,create-index,check-cluster-health,index,refresh-after-index"
```

Previous commit `7b3dd30`:

```
esrally --pipeline="from-sources-complete" --track="nyc_taxis" --car="4gheap" --runtime-jdk=8 --challenge="append-no-conflicts" --revision="7b3dd30" --include-tasks="delete-index,create-index,check-cluster-health,index,refresh-after-index"
```

Note: If you are using Oracle JDK you can also get a profile by specifying `--telemetry=jfr`.

### PR DESCRIPTION
**Pull request by spinscale**
The nightly benchmarks showed a sharp decrease in aggregation performance for
the UTC case.

This commit uses the same calculation as joda time, which requires no
conversion into any java time object, also, the check for an fixedoffset
has been put into the ctor to reduce the need for runtime calculations.
The same goes for the amount of the used unit in milliseconds.

Closes #37826

Benchmark comparison (on my notebook):

Before:

```
RoundingBenchmark.timeUnitRoundingUtcMonthOfYearJava      avgt   30  140,989 ± 14,582  ns/op
RoundingBenchmark.timeUnitRoundingUtcMonthOfYearJoda      avgt   30   11,981 ±  0,066  ns/op
```

final results comparing against the joda implementations (all benchmarks taken on my osx notebook, so take with a grain of salt)

```
Benchmark                                                 Mode  Cnt    Score    Error  Units
RoundingBenchmark.timeIntervalRoundingJava                avgt   30  226,556 ±  1,931  ns/op
RoundingBenchmark.timeIntervalRoundingJoda                avgt   30  328,073 ±  2,480  ns/op
RoundingBenchmark.timeRoundingDateTimeUnitDayOfMonthJava  avgt   30   88,096 ±  1,887  ns/op
RoundingBenchmark.timeRoundingDateTimeUnitDayOfMonthJoda  avgt   30  358,745 ± 22,062  ns/op
RoundingBenchmark.timeRoundingDateTimeUnitJava            avgt   30  133,689 ±  0,774  ns/op
RoundingBenchmark.timeRoundingDateTimeUnitJoda            avgt   30  357,008 ± 19,957  ns/op
RoundingBenchmark.timeUnitRoundingUtcDayOfMonthJava       avgt   30   11,695 ±  0,040  ns/op
RoundingBenchmark.timeUnitRoundingUtcDayOfMonthJoda       avgt   30   11,905 ±  0,056  ns/op
RoundingBenchmark.timeUnitRoundingUtcMonthOfYearJava      avgt   30    3,189 ±  0,021  ns/op
RoundingBenchmark.timeUnitRoundingUtcMonthOfYearJoda      avgt   30    5,465 ±  0,026  ns/op
RoundingBenchmark.timeUnitRoundingUtcQuarterOfYearJava    avgt   30    4,509 ±  0,059  ns/op
RoundingBenchmark.timeUnitRoundingUtcQuarterOfYearJoda    avgt   30   31,511 ±  0,414  ns/op
RoundingBenchmark.timeUnitRoundingUtcYearOfCenturyJava    avgt   30    2,991 ±  0,010  ns/op
RoundingBenchmark.timeUnitRoundingUtcYearOfCenturyJoda    avgt   30    6,456 ±  0,040  ns/op
```

I also added a duelling test class that runs against the old implementation to see if there are any inconsistencies. I ran this class a couple of million times on my linux without failure.

### CODE REVIEW

**Review by danielmitterdorfer — COMMENTED** [review]
I left a few comments. The macrobenchmark results look promising indeed. I did not check the parts that you took from Joda time but I wonder whether we need to attribute this more clearly.

File: server/src/main/java/org/elasticsearch/common/time/DateUtils.java [hunk 1]
```diff
@@ -77,4 +78,224 @@ public static ZoneId of(String zoneId) {
...
+
+    /**
+     * Round down to the beginning of the quarter of the year of the specified time
+     * @param utcMillis the milliseconds since the epoch
+     * @return The milliseconds since the epoch rounded down to the quarter of the year
+     */
+    public static long roundQuarterOfYear(long utcMillis) {
+        int year = DateUtils.getYear(utcMillis);
+        int month = DateUtils.getMonthOfYear(utcMillis, year);
+        return DateUtils.of(year, Month.of(month).firstMonthOfQuarter().getValue());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/time/DateUtils.java]: It looks like this constructs a `Month` object simply to round it down to a multiple of 3. I don't think it loses much clarity to do this by hand instead.

File: server/src/main/java/org/elasticsearch/common/time/DateUtils.java [hunk 2]
```diff
@@ -77,4 +78,224 @@ public static ZoneId of(String zoneId) {
...
+    }
+
+    /**
+     * calculates the first day of a year in milliseconds since the epoch (assuming UTC)
+     *
+     * @param year the year
+     * @return the milliseconds since the epoch of the first of january at midnight of the specified year
+     */
+    // see org.joda.time.chrono.GregorianChronology
+    private static long calculateFirstDayOfYearMillis(int year) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/time/DateUtils.java]: I find the mention of `Day` in this method's name to be a little confusing. Maybe `utcMillisAtStartOfYear`?

**Review by DaveCTurner — COMMENTED** [review]
I left mostly suggestions about the names of things. I think we should have more thorough tests for the copied code, particularly around the boundaries (e.g. month boundaries ±1ms) and with negative numbers.

File: server/src/main/java/org/elasticsearch/common/time/DateUtilsRounding.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    // see org.joda.time.chrono.BasicGJChronology
+    static long getTotalMillisByYearMonth(final int year, final int month) {
+        if (isLeapYear(year)) {
+            return MAX_TOTAL_MILLIS_BY_MONTH_ARRAY[month - 1];
+        } else {
+            return MIN_TOTAL_MILLIS_BY_MONTH_ARRAY[month - 1];
+        }
+    }
+
+    private static final long[] MIN_TOTAL_MILLIS_BY_MONTH_ARRAY;
```
Comment (danielmitterdorfer) [inline server/src/main/java/org/elasticsearch/common/time/DateUtilsRounding.java]: Should we move the static fields and the static initializer to the top of the class?

**Review by danielmitterdorfer — COMMENTED** [review]
Looks mostly fine to me, I have only a remark about the order of fields within one of the classes.

**Review by DaveCTurner — APPROVED** [review]
LGTM. I wrote the following test to satisfy myself that `DateUtilsRounding` does the right thing:

```

    public void testDateUtilsRounding() {
        for (int year = -1000; year < 3000; year++) {
            final long startOfYear = DateUtilsRounding.utcMillisAtStartOfYear(year);
            assertThat(startOfYear, equalTo(ZonedDateTime.of(year, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC).toInstant().toEpochMilli()));

            assertThat(DateUtilsRounding.getYear(startOfYear), equalTo(year));
            assertThat(DateUtilsRounding.getYear(startOfYear - 1), equalTo(year - 1));

            assertThat(DateUtilsRounding.getMonthOfYear(startOfYear, year), equalTo(1));
            assertThat(DateUtilsRounding.getMonthOfYear(startOfYear - 1, year - 1), equalTo(12));

            for (int month = 1; month <= 12; month++) {
                final long startOfMonth = ZonedDateTime.of(year, month, 1, 0, 0, 0, 0, ZoneOffset.UTC).toInstant().toEpochMilli();
                assertThat(DateUtilsRounding.getMonthOfYear(startOfMonth, year), equalTo(month));

                if (month > 1) {
                    assertThat(DateUtilsRounding.getYear(startOfMonth - 1), equalTo(year));
                    assertThat(DateUtilsRounding.getMonthOfYear(startOfMonth - 1, year), equalTo(month - 1));
                }
            }
        }
    }
```

Not sure if you think it's worth adding this to the suite.

**Comment by spinscale** [comment]
@elasticmachine retest this please

**Comment by jpountz** [comment]
@spinscale  I'm assuming there is nothing left to backport and removed the backport pending label.

**Pull request merged**